climakitae/new_core/processors/metric_calc.py

```diff
@@ -0,0 +1,342 @@
+
+        # Calculate percentiles if requested
+        results = []
+        if self.percentiles is not UNSET:
+            percentile_result = data.quantile(
+                [p / PERCENTILE_TO_QUANTILE_FACTOR for p in self.percentiles],
+                dim=calc_dim,
+                keep_attrs=True,
+                skipna=self.skipna,
+            )
+
+            # Rename quantile coordinate to percentiles for clarity
+            if "quantile" in percentile_result.coords:
+                percentile_result = percentile_result.rename({"quantile": "percentile"})
+                # Update coordinate values to be the original percentile values
+                percentile_result = percentile_result.assign_coords(
+                    percentile=self.percentiles
+                )
+
+            results.append(percentile_result)
+
+        # Calculate single metric (unless percentiles_only is True)
+        if not self.percentiles_only:
+            metric_functions = {
+                "min": lambda x: x.min(
+                    dim=calc_dim, keep_attrs=True, skipna=self.skipna
+                ),
+                "max": lambda x: x.max(
+                    dim=calc_dim, keep_attrs=True, skipna=self.skipna
+                ),
+                "mean": lambda x: x.mean(
+                    dim=calc_dim, keep_attrs=True, skipna=self.skipna
+                ),
+                "median": lambda x: x.median(
+                    dim=calc_dim, keep_attrs=True, skipna=self.skipna
+                ),
+            }
+
+            metric_result = metric_functions[self.metric](data)
+            results.append(metric_result)
+
+        # Return combined results or single result
+        if len(results) == 1:
+            return results[0]
+        elif len(results) == 2 and self.percentiles is not None:
+            # Combine percentiles and metric results
+            percentile_result, metric_result = results
+
+            # Create a combined dataset/dataarray
+            if isinstance(data, xr.Dataset):
+                # For datasets, we need to be more careful about combining
+                combined_data = {}
+
+                # Add percentile results
+                for var_name in percentile_result.data_vars:
+                    for i, p in enumerate(self.percentiles):
+                        # Drop the percentile coordinate to avoid conflicts when combining
+                        percentile_data = (
+                            percentile_result[var_name]
+                            .isel(percentile=i)
+                            .drop_vars("percentile")
+                        )
+                        combined_data[f"{var_name}_p{p}"] = percentile_data
+
+                # Add metric results
+                for var_name in metric_result.data_vars:
+                    combined_data[f"{var_name}_{self.metric}"] = metric_result[var_name]
+
+                result = xr.Dataset(combined_data, attrs=data.attrs)
+
+            else:
```

Comment:
This clause, as well as the one above could probably be extracted into their own helper functions.